### PR TITLE
Travin Maksim / Lab 1 / var 2

### PR DESCRIPTION
**Pull request by TMworking**
Add always inlite attribute to functions without conditions

### CODE REVIEW

**Comment by TMworking** [comment]
Pull request was reopened with correction tasks

File: clang/test/lab1/travin_maksim/test.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]|[0-9]+), (line|col):([0-9]+:[0-9]|[0-9]+)> (line|col):([0-9]+:[0-9]|[0-9]+) test1 'void \(\)'}}
+void test1() {};
+// CHECK-NOT: `-AlwaysInlineAttr {{0[xX][0-9a-fA-F]+ <(line|col):([0-9]+:[0-9]|[0-9]+)> Implicit always_inline}}
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]+), line:([0-9]+:[0-9]+)> line:([0-9]+:[0-9]+) test2 'int \(int\)'}}
+int test2(int a) {
+	return a;
+}
+// CHECK-NOT: `-AlwaysInlineAttr {{0[xX][0-9a-fA-F]+ <(line|col):([0-9]+:[0-9]|[0-9]+)> Implicit always_inline}}
```
Comment (m-ly4) [inline clang/test/lab1/travin_maksim/test.cpp]: Why not?

File: clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+public:
+  explicit InlineVisitor(clang::ASTContext *Context)
+      : Context(Context), HasStatement(false) {}
+
+  bool VisitFunc(clang::FunctionDecl *Func) {
+    HasStatement = false;
+    clang::Stmt *BodyFunc = Func->getBody();
+
+    if (!Func->hasAttr<clang::AlwaysInlineAttr>()) {
+      for (const clang::Stmt *statements : BodyFunc->children()) {
```
Comment (m-ly4) [inline clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp]:
```suggestion
    if (Func->hasAttr<clang::AlwaysInlineAttr>()) {
      return true;
    }

    for (const clang::Stmt *statements : BodyFunc->children()) {
```

File: clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        if (clang::isa<clang::IfStmt>(statements) ||
+            clang::isa<clang::WhileStmt>(statements) ||
+            clang::isa<clang::ForStmt>(statements) ||
+            clang::isa<clang::DoStmt>(statements) ||
+            clang::isa<clang::SwitchStmt>(statements)) {
+          HasStatement = true;
```
Comment (m-ly4) [inline clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp]: If we found a conditional statement then we may skip check of remaining nodes.

File: clang/test/lab1/travin_maksim/test.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]+), line:([0-9]+:[0-9]+)> line:([0-9]+:[0-9]+) test0 'void \(int, int\)'}}
+// CHECK: AlwaysInlineAttr {{0[xX][0-9a-fA-F]+ <line:5:21> always_inline}}
+void __attribute__((always_inline)) test0(int a, int b) {
+    a = b;
+}
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]|[0-9]+), (line|col):([0-9]+:[0-9]|[0-9]+)> (line|col):([0-9]+:[0-9]|[0-9]+) test1 'void \(\)'}}
+void test1() {};
+// CHECK-NOT: `-AlwaysInlineAttr {{0[xX][0-9a-fA-F]+ <(line|col):([0-9]+:[0-9]|[0-9]+)> Implicit always_inline}}
```
Comment (m-ly4) [inline clang/test/lab1/travin_maksim/test.cpp]: Why not also?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/test/lab1/travin_maksim/test.cpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -ast-dump -ast-dump-filter test -load %llvmshlibdir/TravinAlwaysInlinePlugin%pluginext -add-plugin add-always-inline %s | FileCheck %s
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]+), line:([0-9]+:[0-9]+)> line:([0-9]+:[0-9]+) test0 'void \(int, int\)'}}
+// CHECK: AlwaysInlineAttr {{0[xX][0-9a-fA-F]+ <line:5:21> always_inline}}
+void __attribute__((always_inline)) test0(int a, int b) {
+    a = b;
+}
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]|[0-9]+), (line|col):([0-9]+:[0-9]|[0-9]+)> (line|col):([0-9]+:[0-9]|[0-9]+) test1 'void \(\)'}}
+void test1();
```
Comment (m-ly4) [inline clang/test/lab1/travin_maksim/test.cpp:11]: Please add a check of presence/absence of attribute.

File: clang/test/lab1/travin_maksim/test.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+// CHECK: FunctionDecl {{0[xX][0-9a-fA-F]+ <.+test\.cpp:([0-9]+:[0-9]+), line:([0-9]+:[0-9]+)> line:([0-9]+:[0-9]+) test2 'int \(int\)'}}
+int test2(int a) {
```
Comment (m-ly4) [inline clang/test/lab1/travin_maksim/test.cpp:15]: The same

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+public:
+  explicit InlineVisitor(clang::ASTContext *Context)
+      : Context(Context), HasStatement(false) {}
+
+  bool VisitFunctionDecl(clang::FunctionDecl *Func) {
+    HasStatement = false;
+    TraverseStmt(Func->getBody());
+    if (Func->hasAttr<clang::AlwaysInlineAttr>()) {
+      Func->dump();
```
Comment (m-ly4) [inline clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp]: Is it necessary? If it is already there then `-ast-dump` option should print it anyway.

Comment (TMworking) [reply clang/labs/lab1/travin_maksim/TravinAlwaysInline.cpp]: No, it's not necessary

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**